# Project 2:  Lucid Titanic Sleuthing

## Part 1: Developing an understanding of the data

#### Based on the description of the data you read in the readme describe in your own words this data.

##### The data sets are for Titanic and Lusitania passenger list and records who died and who survived, along with that other information is provided that may help in performing descriptive statistics. 

#### Based on our conceptual understanding of the columns in this data set.  What are the reasonable range of the values for the Sex, Age, SibSp, Parch columns.

##### For titanic 25% to 75% of passenger age is betwwen 22 and 35.  and for Lusitania it is 29 to 36.
##### for titanic Sibsp has  a mid range of from 0 to 1, parch mid range is zero. Out of 1309 people in titanic majority are male.

In [630]:
titanic_data.describe()

pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1309.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881138     0.498854     0.385027   
std       0.837836     0.486055    12.883193     1.041658     0.865560   
min       1.000000     0.000000     0.170000     0.000000     0.000000   
25%       2.000000     0.000000    22.000000     0.000000     0.000000   
50%       3.000000     0.000000    29.881138     0.000000     0.000000   
75%       3.000000     1.000000    35.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare        body  
count  1308.000000  121.000000  
mean     33.295479  160.809917  
std      51.758668   97.696922  
min       0.000000    1.000000  
25%       7.895800   72.000000  
50%      14.454200  155.000000  
75%      31.275000  256.000000  
max     512.329200  328.000000

In [631]:
lusitania_data.describe()

Age   Value
count  1957.000000  1957.0
mean     32.502887     1.0
std      11.555803     0.0
min       1.000000     1.0
25%      29.000000     1.0
50%      32.502887     1.0
75%      36.000000     1.0
max      76.000000     1.0

### Open the data in sublime text is there anything that jumps out to you?

#### The Age column in lusitania data set has  dirty data that is string and numeric mixed togather and lot of null and ?. The titanic data set has a better quality data compare to lusatina data set. 

## Part 2: reading the data in

#### Now read the data into a Pandas DataFrame

In [544]:
!pwd

/home/jovyan/work/dsi/dsi-workspace/project-02


In [583]:
import pandas as pd
import numpy as np
titanic_data_location = './assets/titanic.csv'
lusitania_data_location = './assets/lusitania.csv'

titanic_data = pd.read_csv(titanic_data_location)
lusitania_data = pd.read_csv(lusitania_data_location)


In [197]:
titanic_data.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [584]:
lusitania_data.head()

Family name Title Personal name   Fate Age Department/Class Passenger/Crew  \
0     CAMERON   Mr.    Charles W.   Lost  38             Band           Crew   
1  CARR-JONES   Mr.            E.   Lost  37             Band           Crew   
2   DRAKEFORD   Mr.        Edward  Saved  30             Band           Crew   
3     HAWKINS   Mr.        Handel  Saved  25             Band           Crew   
4   HEMINGWAY   Mr.  John William  Saved  27             Band           Crew   

  Citizenship     Position Status  ...  Country Lifeboat Rescue Vessel  \
0     British          NaN    NaN  ...      NaN      NaN           NaN   
1     British          NaN    NaN  ...      NaN      NaN           NaN   
2     British       Violin    NaN  ...      NaN      NaN           NaN   
3     British        Cello    NaN  ...      NaN      NaN           NaN   
4     British  Double Bass    NaN  ...      NaN      NaN           NaN   

  Body No. Ticket No. Cabin No. Traveling Companions and other notes Value  \
0      NaN        NaN       NaN                                  NaN     1   
1      NaN        NaN       NaN                                  NaN     1   
2      NaN        NaN       NaN                                  NaN     1   
3      NaN        NaN       NaN                                  NaN     1   
4      NaN        NaN       NaN                                  NaN     1   

  Adult/Minor   Sex  
0       Adult  Male  
1       Adult  Male  
2       Adult  Male  
3       Adult  Male  
4       Adult  Male  

[5 rows x 23 columns]

#### Check that the age column doesn't have any unreasonable values 

# Check Age for both the data sets

Lusitania Data Set

There are non-numeric values e.g. 2-months, 10-months, ?, Infant. Statistics on those cannot be performed. 

In [585]:
pv_age = pd.pivot_table(lusitania_data, columns='Age', aggfunc=[np.sum])
pv_age.head(10)

sum                                                              \
Age   02-months 03-months 05-months 06-months 08-months 09-months  1 1.5 10   
Value         1         4         3         2         2         1  1   2  9   

                 ...                                      
Age   10-months  ...   69  7 70 72 73 76  8  9  ? Infant  
Value         4  ...    1  8  3  1  1  1  8  5  1      8  

[1 rows x 120 columns]

In [586]:
lusitania_data['Age'].unique()

array(['38', '37', '30', '25', '27', '48', nan, '24', '19 ?', '57', '50',
       '56', '41', '19', '33', '29', '18', '20', '21', '26', '17', '58',
       '47', '54', '35', '43', '59', '53', '44', '51', '40', '49', '42',
       '32', '31', '34', '22', '45', '36', '29 ?', '52', '23', '60', '28',
       '16', '46', '15', '39', '63 ?', '55', '64', '53 ?', '09-months',
       '6', '9', '14', '10', '12', '?', '62', '5', '8', '65', '68', '76',
       '61', '63', '03-months', '1.5', '2.5', '18-months', '3', '2',
       '25 ?', '08-months', '4', '15-months', '14-months', 'Infant',
       '3-6-months ?', '51 or 47', '45 or 51', '22 (?)', '17-months',
       '13-months', '02-months', '49 ?', '05-months', '11', '59 (64?)',
       '63 (51?)', '37 (39?)', '32 (22?)', '31 (32?)', '24 (23?)', '27?',
       '7', '42?', '31 ?', '10-months', '38 (?)', '70', '62 (?)',
       '06-months', '13', '2_', '11-months', '30 ?', '34?', '1',
       '39 (47?)', '67', '73', '72', '11-months (3?)', '4.25', '69', '26?'

In [587]:
# clean Age data from lusitania data set

pattern_brackets_1 = '\(..\?\)'
pattern_brackets_2 = '\(.\?\)'
pattern_q = ' \(\?\)'
pattern_q2 = '\?'
pattern_months = '..-months'
pattern_infant = 'Infant'
pattern_spaces = ' '
pattern_underscore = '_'
pattern_exception1 = 'or..'

lusitania_data['Age']=lusitania_data['Age'].str.replace(pattern_brackets_1,'')
lusitania_data['Age']=lusitania_data['Age'].str.replace(pattern_brackets_2,'')
lusitania_data['Age']=lusitania_data['Age'].str.replace(pattern_q,'')
lusitania_data['Age']=lusitania_data['Age'].str.replace(pattern_q2,'')
lusitania_data['Age']=lusitania_data['Age'].str.replace(pattern_months,'1')
lusitania_data['Age']=lusitania_data['Age'].str.replace(pattern_infant,'')
lusitania_data['Age']=lusitania_data['Age'].str.replace(pattern_spaces,'')
lusitania_data['Age']=lusitania_data['Age'].str.replace(pattern_underscore,'')
lusitania_data['Age']=lusitania_data['Age'].str.replace(pattern_exception1,'')
lusitania_data['Age'].unique()

array(['38', '37', '30', '25', '27', '48', nan, '24', '19', '57', '50',
       '56', '41', '33', '29', '18', '20', '21', '26', '17', '58', '47',
       '54', '35', '43', '59', '53', '44', '51', '40', '49', '42', '32',
       '31', '34', '22', '45', '36', '52', '23', '60', '28', '16', '46',
       '15', '39', '63', '55', '64', '1', '6', '9', '14', '10', '12', '',
       '62', '5', '8', '65', '68', '76', '61', '1.5', '2.5', '3', '2', '4',
       '11', '7', '70', '13', '67', '73', '72', '4.25', '69'], dtype=object)

In [588]:
# convert age column to numeric
lusitania_data['Age'] = pd.to_numeric(lusitania_data['Age'])

In [589]:
lusitania_data[pd.isnull(lusitania_data['Age'])]

Family name       Title     Personal name   Fate  Age Department/Class  \
6       ANDERSON         Mr.              John   Lost  NaN             Deck   
7         BATTLE         Mr.             James  Saved  NaN             Deck   
12       CLINTON         Mr.            George  Saved  NaN             Deck   
13         CLYDE         Mr.              Neil  Saved  NaN             Deck   
14         CRANK         Mr.              John   Lost  NaN             Deck   
15    CUNNINGHAM         Mr.            Edward  Saved  NaN             Deck   
16        DAVIES         Mr.              John  Saved  NaN             Deck   
17          EGAN         Mr.           William  Saved  NaN             Deck   
18        ELLIOT         Mr.     Arthur Graham   Lost  NaN             Deck   
19        ELLIOT         Mr.              Fred  Saved  NaN             Deck   
20         EVANS         Mr.    Thomas William   Lost  NaN             Deck   
21         FAYLE         Mr.            Edward  Saved  NaN             Deck   
22      FLEMMING         Mr.           Herbert   Lost  NaN             Deck   
24          GUNN         Mr.                C.  Saved  NaN             Deck   
26       HEFFORD         Mr.             Percy   Lost  NaN             Deck   
28      HENNESEY         Mr.             Frank  Saved  NaN             Deck   
29        HERMAN         Mr.            Albert   Lost  NaN             Deck   
30         HEYES         Mr.                J.  Saved  NaN             Deck   
31        HUGHES         Mr.       John Edward   Lost  NaN             Deck   
33      JOHNSTON         Mr.       Hugh Robert  Saved  NaN             Deck   
35         JONES         Mr.           William   Lost  NaN             Deck   
36        KNIGHT         Mr.             C. T.  Saved  NaN             Deck   
39       LOCKETT         Mr.                H.  Saved  NaN             Deck   
40        MALONE         Mr.                F.   Lost  NaN             Deck   
42       McGLADE         Mr.            George   Lost  NaN             Deck   
43        McLEOD         Mr.            Victor   Lost  NaN             Deck   
44       McNALLY         Mr.             Peter   Lost  NaN             Deck   
45         MOORE         Mr.            Daniel  Saved  NaN             Deck   
49       O'NEILL         Mr.             F. H.  Saved  NaN             Deck   
50         OWENS         Mr.              Owen   Lost  NaN             Deck   
...          ...         ...               ...    ...  ...              ...   
1921    SPENDLEY        Mr.        David March   Lost  NaN      Victualling   
1922     STANLEY        Mr.           David H.  Saved  NaN      Victualling   
1923  STANSFIELD        Mr.             Leslie   Lost  NaN      Victualling   
1924       STARK        Mr.       Robert Lucas   Lost  NaN      Victualling   
1925   STEINBERG  Master (?)  Arthur Valentine   Lost  NaN      Victualling   
1928      SUTTAN        Mr.       Nugent Moore   Lost  NaN      Victualling   
1929      SUTTON        Mr.              Frank  Saved  NaN      Victualling   
1930       SWIFT        Mr.             Edward   Lost  NaN      Victualling   
1932      THOMAS        Mr.       George Henry   Lost  NaN      Victualling   
1933      THOMAS        Mr.               John  Saved  NaN      Victualling   
1934       THORN        Mr.     Alfred Richard   Lost  NaN      Victualling   
1935    THORNLEY        Mr.             George  Saved  NaN      Victualling   
1936       ULGAR        Mr.               Jean   Lost  NaN      Victualling   
1938      VARNEY        Mr.               John   Lost  NaN      Victualling   
1939      VERDIN        Mr.            Francis   Lost  NaN      Victualling   
1940     WALLACE        Mr.              James  Saved  NaN      Victualling   
1941       WEIGH        Mr.            Edward�  Saved  NaN      Victualling   
1943     WHALLEY        Mr.             Leslie  Saved  NaN      Victualling   
1945   WILKINSON        Mr.          Mr. James   Lost  

In [590]:
# get the mean age for respective group and update all null with that value
a=lusitania_data.groupby(['Sex','Adult/Minor'])['Age'].mean()
a

Sex     Adult/Minor
Female  Adult          35.621984
        Infant          3.541667
        Minor           7.840000
Male    Adult          36.211964
        Infant          1.019231
        Minor           9.076087
        Minor?               NaN
Name: Age, dtype: float64

In [591]:
lusitania_mean_age = lusitania_data.pivot_table('Age',
                                      columns=['Adult/Minor'],
                                      index=['Sex'],
                                      aggfunc=np.mean)

In [592]:
lusitania_mean_age.query('Sex == ["Female"]')['Adult']
#titanic_data_summary_by_sex_status.query(Sex == ["Female"]')

Sex
Female    35.621984
Name: Adult, dtype: float64

In [593]:
np.mean(lusitania_data['Age'])

32.502886836027713

In [ ]:
# update null values in Female rows with female mean
# need to work on this.
#lusitania_data['Age'] = lusitania_data['Age']\
#            .apply(lambda x: 'Third' if x == 'Third (Distressed British Seaman)' else x) 

In [594]:
# replace all nulls with mean for both the data sets
(lusitania_data['Age'].fillna(np.mean(lusitania_data['Age']), inplace=True))

In [595]:
# replace all nulls with mean
(titanic_data['age'].fillna(np.mean(titanic_data['age']), inplace=True))

In [596]:
titanic_data[pd.isnull(titanic_data['age'])]

Empty DataFrame
Columns: [pclass, survived, name, sex, age, sibsp, parch, ticket, fare, cabin, embarked, boat, body, home.dest]
Index: []

In [597]:
lusitania_data[pd.isnull(lusitania_data['Age'])]

Empty DataFrame
Columns: [Family name, Title, Personal name, Fate, Age, Department/Class, Passenger/Crew, Citizenship, Position, Status, City, County, State, Country, Lifeboat, Rescue Vessel, Body No., Ticket No., Cabin No., Traveling Companions and other notes, Value, Adult/Minor, Sex]
Index: []

[0 rows x 23 columns]

# Check department/class column

In [598]:
## check unique values for department/class column
lusitania_data.groupby(['Department/Class'])['Department/Class'].count()

Department/Class
Band                                   5
Deck                                  69
Engineering                          313
Saloon                               291
Second                               601
Stowaway                               3
Third                                370
Third (Distressed British Seaman)      3
Victualling                          306
Name: Department/Class, dtype: int64

In [599]:
# update 'Third (Distressed British Seaman)' with 'Third' as there are only three occurance and they are passengers 
#econ_data_df.drop('x_bar_tmp', axis=1, inplace=True)
lusitania_data['Department/Class'] = lusitania_data['Department/Class']\
            .apply(lambda x: 'Third' if x == 'Third (Distressed British Seaman)' else x) 

In [600]:
## check unique values for department/class column
lusitania_data.groupby(['Department/Class'])['Department/Class'].count()

Department/Class
Band             5
Deck            69
Engineering    313
Saloon         291
Second         601
Stowaway         3
Third          373
Victualling    306
Name: Department/Class, dtype: int64

In [601]:
# check passenger/crew  column unique values and count
lusitania_data.groupby(['Passenger/Crew'])['Passenger/Crew'].count()

Passenger/Crew
Crew          693
Passenger    1265
Stowaway        3
Name: Passenger/Crew, dtype: int64

In [602]:
# remove stowaway from passenger/crew column as are exception in this case. Although important to note that they are 
#human being. #the anaalysis is focused on pasengers
lusitania_data = lusitania_data[lusitania_data['Passenger/Crew'] != 'Stowaway']


In [603]:
# check passenger/crew  column unique values and count
lusitania_data.groupby(['Passenger/Crew'])['Passenger/Crew'].count()

Passenger/Crew
Crew          693
Passenger    1265
Name: Passenger/Crew, dtype: int64

In [604]:
# check Fate column unique values and count
lusitania_data.groupby(['Fate'])['Fate'].count()

Fate
Lost                        1190
Not on board                   1
Saved                        763
Saved (died from trauma)       4
Name: Fate, dtype: int64

In [605]:
# remove 'not on board' from Fate as it is not relevant for analysis. 
lusitania_data = lusitania_data[lusitania_data['Fate'] != 'Not on board']

In [606]:
# check Fate column unique values and count
lusitania_data.groupby(['Fate'])['Fate'].count()

Fate
Lost                        1190
Saved                        763
Saved (died from trauma)       4
Name: Fate, dtype: int64

In [607]:
## check unique values for adult/minor column
lusitania_data.groupby(['Adult/Minor'])['Adult/Minor'].count()

Adult/Minor
Adult     1787
Infant      47
Minor      119
Minor?       4
Name: Adult/Minor, dtype: int64

In [608]:
# update 'Minor?' with 'Minor' as there are only 4 occurance and they are passengers 
lusitania_data['Adult/Minor'] = lusitania_data['Adult/Minor']\
            .apply(lambda x: 'Minor' if x == 'Minor?' else x) 

In [609]:
## check unique values for adult/minor column
lusitania_data.groupby(['Adult/Minor'])['Adult/Minor'].count()

Adult/Minor
Adult     1787
Infant      47
Minor      123
Name: Adult/Minor, dtype: int64

In [610]:
#create a column for deceased for the descriptive statistics
#titanic_data['deceased'] = titanic_data['survived'].apply(lambda x: 1 if x == 0 else 0)
#titanic_data['Value']=1
titanic_data['status'] = titanic_data['survived'].apply(lambda x: 'deceased' if x == 0 else 'survived')
#titanic_data["status"] = titanic_data["status"].astype("category")
#titanic_data["status"].cat.set_categories(["survived","deceased"],inplace=True)
labels = [ "{0} - {1}".format(i, i + 9) for i in range(0, 100, 10) ]
titanic_data['age_group'] = pd.cut(titanic_data['age'], range(0, 105, 10), right=False, labels=labels)

In [611]:
titanic_data

pclass  survived                                               name  \
0          1         1                      Allen, Miss. Elisabeth Walton   
1          1         1                     Allison, Master. Hudson Trevor   
2          1         0                       Allison, Miss. Helen Loraine   
3          1         0               Allison, Mr. Hudson Joshua Creighton   
4          1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5          1         1                                Anderson, Mr. Harry   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
7          1         0                             Andrews, Mr. Thomas Jr   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9          1         0                            Artagaveytia, Mr. Ramon   
10         1         0                             Astor, Col. John Jacob   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12         1         1                      Aubart, Mme. Leontine Pauline   
13         1         1                       Barber, Miss. Ellen "Nellie"   
14         1         1               Barkworth, Mr. Algernon Henry Wilson   
15         1         0                                Baumann, Mr. John D   
16         1         0                           Baxter, Mr. Quigg Edmond   
17         1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18         1         1                              Bazzani, Miss. Albina   
19         1         0                               Beattie, Mr. Thomson   
20         1         1                      Beckwith, Mr. Richard Leonard   
21         1         1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   
22         1         1                              Behr, Mr. Karl Howell   
23         1         1                              Bidois, Miss. Rosalie   
24         1         1                                  Bird, Miss. Ellen   
25         1         0                                Birnbaum, Mr. Jakob   
26         1         1                            Bishop, Mr. Dickinson H   
27         1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
28         1         1                             Bissette, Miss. Amelia   
29         1         1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   
...      ...       ...                                                ...   
1279       3         0               Vestrom, Miss. Hulda Amanda Adolfina   
1280       3         0                                    Vovk, Mr. Janko   
1281       3         0                               Waelens, Mr. Achille   
1282       3         0                                Ware, Mr. Frederick   
1283       3         0                        Warren, Mr. Charles William   
1284       3         0                                  Webber, Mr. James   
1285       3         0                                Wenzel, Mr. Linhart   
1286       3         1    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1287       3         0                   Widegren, Mr. Carl/Charles Peter   
1288       3         0                          Wiklund, Mr. Jakob Alfred   
1289       3         0                            Wiklund, Mr. Karl Johan   
1290       3         1                   Wilkes, Mrs. James (Ellen Needs)   
1291       3         0                   Willer, Mr. Aaron ("Abi Weller")   
1292       3         0                                 Willey, Mr. Edward   
1293       3         0                  Williams, Mr. Howard Hugh "Harry"   
1294       3         0                               Williams, Mr. Leslie   
1295       3         0                                Windelov, Mr. Einar   
1296       3         0                                   Wirz, Mr. Albert   
1297       3         0                             Wiseman, Mr. Phillippe   
1298       3         0                          Wittevrongel, Mr. Camille   
12

#### Check for missing values.  How do you know that a value is missing?

I used isnull function of pandas to look for null/missing values in data sets. Please see above command.

Titanic Data Set

1) Age column have 263 null values that is unreasonable for the purpose of this data. Below command shows that.

Lusitania Data Set

1) Age column has 658 null values which is a missing value. See below. 

In [612]:
titanic_data[pd.isnull(titanic_data['age'])].head(5)


Empty DataFrame
Columns: [pclass, survived, name, sex, age, sibsp, parch, ticket, fare, cabin, embarked, boat, body, home.dest, status]
Index: []

In [613]:
#lusitania_data[pd.isnull(lusitania_data['Age'])].head(5)
lusitania_data[lusitania_data.Age.isnull()]

Empty DataFrame
Columns: [Family name, Title, Personal name, Fate, Age, Department/Class, Passenger/Crew, Citizenship, Position, Status, City, County, State, Country, Lifeboat, Rescue Vessel, Body No., Ticket No., Cabin No., Traveling Companions and other notes, Value, Adult/Minor, Sex]
Index: []

[0 rows x 23 columns]

#### Does it makes sense to guess at the value?

A guess based on the data value is reasonable e.g. some of the Age values in lusitania dataset had 'months' in it so you can translate those values to digit portion of the value. Where as some had ? in it which is not best to guess. 

## Part 3: data imputation

#### Well let’s say that it does... You likely noticed that Age has some missing values. How many are missing?

Titanic Data Set

1) Age column have 263 null values that is unreasonable for the purpose of this data. Below command shows that.

Lusitania Data Set

1) Age column has 658 null values which is a missing value. See below. 

#### For the Age of the passangers ... how would you guess at the missing values using the other data present in the CSV.

   #### There were significant missing values for Age column in both the data sets and so excluding those from the sample will  make the sample even smaller and probably not help with the statistics. 

   #### so what I did was update all missing values with the mean for 33% of data. Here is how I did it:

(lusitania_data['Age'].fillna(np.mean(lusitania_data['Age']), inplace=True))



## Part 4: Group Statistics

In [614]:
import matplotlib.pyplot as plot
%matplotlib inline
import seaborn
"numpy version: " + np.__version__

'numpy version: 1.10.4'

#### Are there any groups that were especially adversely affected in the Titanic wreck? (justify your response numerically)

### I have reviewed differnt groups. (group statistics in pivot tableshow below) 
#### 1) pclass, sex and embarked combined with status (deceased, survived)
#### 2) pclass and status (deceased, survived),
#### 3) age group and status (deceased, survived),
#### 4) sex and status (deceased, survived)
#### 5) sex, age group and status (deceased, survived)



The data revelas more numbers for deceased compare to survived. The number for deceased(809) people is higher than survived(500). 

#### 1) Titanic summary by pclass, sex, embarked and status (deceased, survived)

#### Table below represents a pivot table by pclass, sex, embarked and status


In [616]:
# titanic summary by pclass, sex, embarked and status
titanic_data_summary = titanic_data.pivot_table('age',
                                      columns=['status'],
                                      index=['pclass','sex','embarked'],
                                      aggfunc='count')
titanic_data_summary['All']=\
                                    titanic_data_summary['survived']\
                                    +titanic_data_summary['deceased']
titanic_data_summary['SurvivalRate']=\
                                    titanic_data_summary['survived']\
                                    /titanic_data_summary['All']
titanic_data_summary['DeathRate']=\
                                    titanic_data_summary['deceased']\
                                    /titanic_data_summary['All']
titanic_data_summary

status                  deceased  survived    All  SurvivalRate  DeathRate
pclass sex    embarked                                                    
1      female C              2.0      69.0   71.0      0.971831   0.028169
              Q              NaN       2.0    NaN           NaN        NaN
              S              3.0      66.0   69.0      0.956522   0.043478
       male   C             42.0      28.0   70.0      0.400000   0.600000
              Q              1.0       NaN    NaN           NaN        NaN
              S             75.0      33.0  108.0      0.305556   0.694444
2      female C              NaN      11.0    NaN           NaN        NaN
              Q              NaN       2.0    NaN           NaN        NaN
              S             12.0      81.0   93.0      0.870968   0.129032
       male   C             12.0       5.0   17.0      0.294118   0.705882
              Q              5.0       NaN    NaN           NaN        NaN
              S            129.0      20.0  149.0      0.134228   0.865772
3      female C              9.0      22.0   31.0      0.709677   0.290323
              Q             23.0      33.0   56.0      0.589286   0.410714
              S             78.0      51.0  129.0      0.395349   0.604651
       male   C             55.0      15.0   70.0      0.214286   0.785714
              Q             50.0       7.0   57.0      0.122807   0.877193
              S            313.0      53.0  366.0      0.144809   0.855191

In [617]:
print 'Total Deceased: ', titanic_data_summary['deceased'].sum()
print 'Total Survived:', titanic_data_summary['survived'].sum()

Total Deceased:  809.0
Total Survived: 498.0


#### 2) Titanic summary by pclass and status(deceased, survived)

#### 74.47% of passengers in pclass 3 died compare to 38.08% in pclass 1 and 57.04% in pclass 2. 

In [618]:
# titanic summary by pclass and status
titanic_data_summary_by_class_status = titanic_data.pivot_table('age',
                                      columns=['status'],
                                      index=['pclass'],
                                      aggfunc='count')
titanic_data_summary_by_class_status['All']=\
                                    titanic_data_summary_by_class_status['survived']\
                                    +titanic_data_summary_by_class_status['deceased']
titanic_data_summary_by_class_status['SurvivalRate']=\
                                    (titanic_data_summary_by_class_status['survived']\
                                    /titanic_data_summary_by_class_status['All'])
titanic_data_summary_by_class_status['DeathRate']=\
                                    titanic_data_summary_by_class_status['deceased']\
                                    /titanic_data_summary_by_class_status['All']
titanic_data_summary_by_class_status

status  deceased  survived  All  SurvivalRate  DeathRate
pclass                                                  
1            123       200  323      0.619195   0.380805
2            158       119  277      0.429603   0.570397
3            528       181  709      0.255289   0.744711

#### 3) Titanic summary by Age Group and status

#### The table below shows passengers by age group and that reveals that of the total 1309 passengers 607 passengers are from age group between 20 and 29 years. 407 passengers died in this age group with a death rate of 67.05% which is the third highest among all age group. Lowest death rate is among age grop between infant and age 10. 

In [621]:
# titanic summary by age group and status
titanic_data_summary_by_agegrp_status  = titanic_data.pivot_table('age',
                                      columns=['status'],
                                      index=['age_group'],
                                      aggfunc='count')
titanic_data_summary_by_agegrp_status ['All']=\
                                    titanic_data_summary_by_agegrp_status ['survived']\
                                    +titanic_data_summary_by_agegrp_status ['deceased']
titanic_data_summary_by_agegrp_status ['SurvivalRate']=\
                                    (titanic_data_summary_by_agegrp_status ['survived']\
                                    /titanic_data_summary_by_agegrp_status ['All'])
titanic_data_summary_by_agegrp_status ['DeathRate']=\
                                    (titanic_data_summary_by_agegrp_status ['deceased']\
                                    /titanic_data_summary_by_agegrp_status ['All'])

titanic_data_summary_by_agegrp_status 

status     deceased  survived    All  SurvivalRate  DeathRate
age_group                                                    
0 - 9          32.0      50.0   82.0      0.609756   0.390244
10 - 19        87.0      56.0  143.0      0.391608   0.608392
20 - 29       407.0     200.0  607.0      0.329489   0.670511
30 - 39       134.0      98.0  232.0      0.422414   0.577586
40 - 49        83.0      52.0  135.0      0.385185   0.614815
50 - 59        38.0      32.0   70.0      0.457143   0.542857
60 - 69        22.0      10.0   32.0      0.312500   0.687500
70 - 79         6.0       1.0    7.0      0.142857   0.857143
80 - 89         NaN       1.0    NaN           NaN        NaN
90 - 99         NaN       NaN    NaN           NaN        NaN

#### Are there any groups that outperformed the survival of the latter group? (justify your response numerically)


#### 4) Titanic summary by sex and status (deceased, survived)

#### The survival rate for female is 67.8% and for male it is 32.2%.
#### So comparitively female survival rate is much higher than male survival rate. See below table.

#### table below also shows survial rate among different age group of female is similar. However survival rate for male age more than 10 is very low.

In [622]:
#summary by sex status
titanic_data_summary_by_sex_status = titanic_data.pivot_table('age',
                                      columns=['status'],
                                      index=['sex'],
                                      aggfunc='count')

titanic_data_summary_by_sex_status['All']=\
                                    titanic_data_summary_by_sex_status['survived']\
                                    +titanic_data_summary_by_sex_status['deceased']
titanic_data_summary_by_sex_status['SurvivalRate']=\
                                    titanic_data_summary_by_sex_status['survived']\
                                    /titanic_data_summary_by_sex_status['All']
titanic_data_summary_by_sex_status
#titanic_data_summary_by_sex_status.query('status == ["female"]')

status  deceased  survived  All  SurvivalRate
sex                                          
female       127       339  466      0.727468
male         682       161  843      0.190985

#### 5) Titanic summary by sex, age group and status (deceased, survived)

In [623]:
#summary by sex and age group and status
titanic_data_summary_by_sex_ag_status  = titanic_data.pivot_table('age',
                                      columns=['status'],
                                      index=['sex', 'age_group'],
                                      aggfunc='count')
titanic_data_summary_by_sex_ag_status ['All']=\
                                    titanic_data_summary_by_sex_ag_status ['survived']\
                                    +titanic_data_summary_by_sex_ag_status ['deceased']
titanic_data_summary_by_sex_ag_status ['SurvivalRate']=\
                                    (titanic_data_summary_by_sex_ag_status ['survived']\
                                    /titanic_data_summary_by_sex_ag_status ['All'])
titanic_data_summary_by_sex_ag_status 

status            deceased  survived    All  SurvivalRate
sex    age_group                                         
female 0 - 9          14.0      25.0   39.0      0.641026
       10 - 19        17.0      47.0   64.0      0.734375
       20 - 29        64.0     129.0  193.0      0.668394
       30 - 39        18.0      68.0   86.0      0.790698
       40 - 49        10.0      36.0   46.0      0.782609
       50 - 59         2.0      25.0   27.0      0.925926
       60 - 69         2.0       8.0   10.0      0.800000
       70 - 79         NaN       1.0    NaN           NaN
       80 - 89         NaN       NaN    NaN           NaN
       90 - 99         NaN       NaN    NaN           NaN
male   0 - 9          18.0      25.0   43.0      0.581395
       10 - 19        70.0       9.0   79.0      0.113924
       20 - 29       343.0      71.0  414.0      0.171498
       30 - 39       116.0      30.0  146.0      0.205479
       40 - 49        73.0      16.0   89.0      0.179775
       50 - 59        36.0       7.0   43.0      0.162791
       60 - 69        20.0       2.0   22.0      0.090909
       70 - 79         6.0       NaN    NaN           NaN
       80 - 89         NaN       1.0    NaN           NaN
       90 - 99         NaN       NaN    NaN           NaN

## Part 5:  Comparative Statistics:  Lusitania

#### Are there any groups that were especially adversely affected in the Lusitania wreck? (justify your response numerically)

### I have created several groups to review the adverse affect in Lusitania data set, and they are following:
#### 1) Summary by Sex, Passenger/crew, department and Fate
#### 2) Summary by Sex, department and Fate
#### 3) Summary by Sex and Fate
#### 4) Summary by department and Fate
#### 5) Summary by passenger/crew and Fate

#### All different categories suggest that the survival rate is distributed evenly and there does not seem to be any anamolies.

In [625]:
# Lusitania summary by sex, passenger/crew, department and fate
lusitania_summary_by_sex_pass_dept_fate = lusitania_data.pivot_table('Value',
                                      columns=['Fate'],
                                      index=['Sex','Passenger/Crew','Department/Class'],
                                      aggfunc='count')
lusitania_summary_by_sex_pass_dept_fate['All']=\
                                    lusitania_summary_by_sex_pass_dept_fate['Lost']\
                                    +lusitania_summary_by_sex_pass_dept_fate['Saved']\
                                    +lusitania_summary_by_sex_pass_dept_fate['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x)

lusitania_summary_by_sex_pass_dept_fate['SurvivalRate']=\
                                    lusitania_summary_by_sex_pass_dept_fate['Saved']\
                                    /lusitania_summary_by_sex_pass_dept_fate['All']

lusitania_summary_by_sex_pass_dept_fate['DeathRate']=\
                                    (lusitania_summary_by_sex_pass_dept_fate['Lost']\
                                     +lusitania_summary_by_sex_pass_dept_fate['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x))\
                                    /lusitania_summary_by_sex_pass_dept_fate['All']

lusitania_summary_by_sex_pass_dept_fate

Fate                                     Lost  Saved  \
Sex    Passenger/Crew Department/Class                 
Female Crew           Victualling        16.0    9.0   
       Passenger      Saloon             56.0   34.0   
                      Second            185.0  108.0   
                      Third              68.0   40.0   
Male   Crew           Band                2.0    3.0   
                      Deck               32.0   37.0   
                      Engineering       201.0  112.0   
                      Victualling       151.0  129.0   
       Passenger      Saloon            121.0   78.0   
                      Second            187.0  119.0   
                      Third             171.0   94.0   

Fate                                    Saved (died from trauma)    All  \
Sex    Passenger/Crew Department/Class                                    
Female Crew           Victualling                            NaN   25.0   
       Passenger      Saloon                                 NaN   90.0   
                      Second                                 1.0  294.0   
                      Third                                  NaN  108.0   
Male   Crew           Band                                   NaN    5.0   
                      Deck                                   NaN   69.0   
                      Engineering                            NaN  313.0   
                      Victualling                            1.0  281.0   
       Passenger      Saloon                                 1.0  200.0   
                      Second                                 1.0  307.0   
                      Third                                  NaN  265.0   

Fate                                    SurvivalRate  DeathRate  
Sex    Passenger/Crew Department/Class                           
Female Crew           Victualling           0.360000   0.640000  
       Passenger      Saloon                0.377778   0.622222  
                      Second                0.367347   0.632653  
                      Third                 0.370370   0.629630  
Male   Crew           Band                  0.600000   0.400000  
                      Deck                  0.536232   0.463768  
                      Engineering           0.357827   0.642173  
                      Victualling           0.459075   0.540925  
       Passenger      Saloon                0.390000   0.610000  
                      Second                0.387622   0.612378  
                      Third                 0.354717   0.645283

In [626]:
# Lusitania summary by sex, department and fate
lusitania_summary_by_sex_dept_fate = lusitania_data.pivot_table('Value',
                                      columns=['Fate'],
                                      index=['Sex','Department/Class'],
                                      aggfunc='count')
lusitania_summary_by_sex_dept_fate['All']=\
                                    lusitania_summary_by_sex_dept_fate['Lost']\
                                    +lusitania_summary_by_sex_dept_fate['Saved']\
                                    +lusitania_summary_by_sex_dept_fate['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x)

lusitania_summary_by_sex_dept_fate['SurvivalRate']=\
                                    lusitania_summary_by_sex_dept_fate['Saved']\
                                    /lusitania_summary_by_sex_dept_fate['All']

lusitania_summary_by_sex_dept_fate['DeathRate']=\
                                    (lusitania_summary_by_sex_dept_fate['Lost']\
                                     +lusitania_summary_by_sex_dept_fate['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x))\
                                    /lusitania_summary_by_sex_dept_fate['All']

lusitania_summary_by_sex_dept_fate

Fate                      Lost  Saved  Saved (died from trauma)    All  \
Sex    Department/Class                                                  
Female Saloon             56.0   34.0                       NaN   90.0   
       Second            185.0  108.0                       1.0  294.0   
       Third              68.0   40.0                       NaN  108.0   
       Victualling        16.0    9.0                       NaN   25.0   
Male   Band                2.0    3.0                       NaN    5.0   
       Deck               32.0   37.0                       NaN   69.0   
       Engineering       201.0  112.0                       NaN  313.0   
       Saloon            121.0   78.0                       1.0  200.0   
       Second            187.0  119.0                       1.0  307.0   
       Third             171.0   94.0                       NaN  265.0   
       Victualling       151.0  129.0                       1.0  281.0   

Fate                     SurvivalRate  DeathRate  
Sex    Department/Class                           
Female Saloon                0.377778   0.622222  
       Second                0.367347   0.632653  
       Third                 0.370370   0.629630  
       Victualling           0.360000   0.640000  
Male   Band                  0.600000   0.400000  
       Deck                  0.536232   0.463768  
       Engineering           0.357827   0.642173  
       Saloon                0.390000   0.610000  
       Second                0.387622   0.612378  
       Third                 0.354717   0.645283  
       Victualling           0.459075   0.540925

In [627]:
# Lusitania summary by sex, and fate
lusitania_summary_by_sex_fate = lusitania_data.pivot_table('Value',
                                      columns=['Fate'],
                                      index=['Sex'],
                                      aggfunc='count')
lusitania_summary_by_sex_fate ['All']=\
                                    lusitania_summary_by_sex_fate ['Lost']\
                                    +lusitania_summary_by_sex_fate ['Saved']\
                                    +lusitania_summary_by_sex_fate ['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x)

lusitania_summary_by_sex_fate ['SurvivalRate']=\
                                    lusitania_summary_by_sex_fate ['Saved']\
                                    /lusitania_summary_by_sex_fate ['All']

lusitania_summary_by_sex_fate ['DeathRate']=\
                                    (lusitania_summary_by_sex_fate ['Lost']\
                                     +lusitania_summary_by_sex_fate ['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x))\
                                    /lusitania_summary_by_sex_fate ['All']

lusitania_summary_by_sex_fate 

Fate    Lost  Saved  Saved (died from trauma)   All  SurvivalRate  DeathRate
Sex                                                                         
Female   325    191                         1   517      0.369439   0.630561
Male     865    572                         3  1440      0.397222   0.602778

In [628]:
# Lusitania summary by dept, and fate
lusitania_summary_by_dept_fate  = lusitania_data.pivot_table('Value',
                                      columns=['Fate'],
                                      index=['Department/Class'],
                                      aggfunc='count')
lusitania_summary_by_dept_fate  ['All']=\
                                    lusitania_summary_by_dept_fate  ['Lost']\
                                    +lusitania_summary_by_dept_fate  ['Saved']\
                                    +lusitania_summary_by_dept_fate  ['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x)

lusitania_summary_by_dept_fate  ['SurvivalRate']=\
                                    lusitania_summary_by_dept_fate  ['Saved']\
                                    /lusitania_summary_by_dept_fate  ['All']

lusitania_summary_by_dept_fate  ['DeathRate']=\
                                    (lusitania_summary_by_dept_fate  ['Lost']\
                                     +lusitania_summary_by_dept_fate  ['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x))\
                                    /lusitania_summary_by_dept_fate  ['All']

lusitania_summary_by_dept_fate  

Fate               Lost  Saved  Saved (died from trauma)    All  SurvivalRate  \
Department/Class                                                                
Band                2.0    3.0                       NaN    5.0      0.600000   
Deck               32.0   37.0                       NaN   69.0      0.536232   
Engineering       201.0  112.0                       NaN  313.0      0.357827   
Saloon            177.0  112.0                       1.0  290.0      0.386207   
Second            372.0  227.0                       2.0  601.0      0.377704   
Third             239.0  134.0                       NaN  373.0      0.359249   
Victualling       167.0  138.0                       1.0  306.0      0.450980   

Fate              DeathRate  
Department/Class             
Band               0.400000  
Deck               0.463768  
Engineering        0.642173  
Saloon             0.613793  
Second             0.622296  
Third              0.640751  
Victualling        0.549020

In [629]:
# Lusitania summary by passenger/crew and fate
lusitania_summary_by_pass_fate   = lusitania_data.pivot_table('Value',
                                      columns=['Fate'],
                                      index=['Passenger/Crew'],
                                      aggfunc='count')
lusitania_summary_by_pass_fate   ['All']=\
                                    lusitania_summary_by_pass_fate   ['Lost']\
                                    +lusitania_summary_by_pass_fate   ['Saved']\
                                    +lusitania_summary_by_pass_fate   ['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x)

lusitania_summary_by_pass_fate   ['SurvivalRate']=\
                                    lusitania_summary_by_pass_fate   ['Saved']\
                                    /lusitania_summary_by_pass_fate   ['All']

lusitania_summary_by_pass_fate   ['DeathRate']=\
                                    (lusitania_summary_by_pass_fate   ['Lost']\
                                     +lusitania_summary_by_pass_fate   ['Saved (died from trauma)']\
                                    .apply(lambda x: 0 if pd.isnull(x) else x))\
                                    /lusitania_summary_by_pass_fate   ['All']

lusitania_summary_by_pass_fate   

Fate            Lost  Saved  Saved (died from trauma)   All  SurvivalRate  \
Passenger/Crew                                                              
Crew             402    290                         1   693      0.418470   
Passenger        788    473                         3  1264      0.374209   

Fate            DeathRate  
Passenger/Crew             
Crew             0.581530  
Passenger        0.625791

#### Are there any groups that outperformed the survival of the latter group? (justify your response numerically)



#### Looking at the department and Fate poivot table 'Victualling' categories have the highest survival rate i.e. 45% compare to other cateogories. 

####  Category 'Band'  seems to outperform survival rate in all categories. However the number of obeservations for band is exceptionally low. The survival rate for Band is 60% compare to all other categories.

#### What does the group-wise survival rate imply about circumstances during these two accidents?

#### The survival rate in lusitania seems to suggest comparatively normal distrbution than titanic. Due to lack of time I am unable to write further. 